Add images to tags #806, TCX import enhancements, tour concatenation fixes and enhancements, tour blog enhancement

FJBDev opened the pull request:
- FIT SDK 21.105

- TCX: Import of calories and average power values

- Tour Blog View enhancement: Now displaying the weather summary
![image](https://user-images.githubusercontent.com/25235665/218220290-d7e86ff6-54d6-4406-b7fd-304fe2601575.png)


- Tour concatenation: 
Fixed bugs: markers and waypoints improperly concatenated (added UI tests to verify that)
Improved the main dialog menu (see below)

**BEFORE**:
![image](https://user-images.githubusercontent.com/25235665/218213937-43eb4897-f210-4da9-b63d-b9cc69e1822d.png)

**AFTER**

**- Ability to delete the source tours (This feature needs to be manually unlocked by the user)**

**- If "Keep original start time for each tour" is selected, then the start time is greyed out and a pause is automatically inserted between each tours since they are not continuously concatenated**


![image](https://user-images.githubusercontent.com/25235665/218215062-102e9e0a-2fe1-43d1-ae4f-52673c175ddf.png)


- Tour Tag: Added the ability to add an image to a tag (link to feature https://github.com/mytourbook/mytourbook/issues/805)

- Removed the [Apache HTTP library](https://hc.apache.org/) (saved 1MB)

- Several packages/libraries updated


**Possible improvements:**

~~- When a large tag image is selected, the displayed image is "pixelized" even though the anti-alias is on (@wolfgang-ch Maybe you had this issue before and you know a possible fix ?)~~

![image](https://user-images.githubusercontent.com/25235665/218215849-3c306288-a2e6-45ed-8d2a-a0da5ae83183.png)

**DONE**
![image](https://user-images.githubusercontent.com/25235665/220477880-0e572761-dddf-4a82-9496-bc1b1bc04835.png)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/tag/Dialog_TourTag.java:
```diff
@@ -36,22 +49,39 @@
  */
 public class Dialog_TourTag extends TitleAreaDialog {
```
I opened this dialog with 4 different tags but all images are not disposed. It is very important that images are disposed when not needed anymore.

I strongly recommend to use the Sleak view to verify the image resources !

![image](https://user-images.githubusercontent.com/1283445/218257769-06fbe292-1ac2-4e1d-9cb2-ccdb7512fff9.png)


I stop for now the review until this is fixed. In the following graph you can see the files which I've not yet reviewed

![image](https://user-images.githubusercontent.com/1283445/218257897-12614fc0-f779-437b-a81f-1ce0356224d2.png)

FJBDev replied:
> I stop for now the review until this is fixed. In the following graph you can see the files which I've not yet reviewed
> 

That sounds good. I am putting the PR in draft mode until I fix all your suggestions/reviews and it is ready to be reviewed again

FJBDev replied:
I already use Sleak but it has never shown any images non disposed when I developed this feature

I used [your documentation ](https://mytourbook.sourceforge.io/mytourbook/index.php/development/documentation-dev-tips/dev-tips-sleak)but maybe I am missing something in my Eclipse configuration ?

![image](https://user-images.githubusercontent.com/25235665/218578597-769dc402-04fc-4138-b28a-d6d94e32f55c.png)

FJBDev replied:
Ok, thanks for the detailed steps, I think I was forgetting to do that :

Press `Snap & Diff` button in Sleak view, important, press it **BEFORE** opening the pref dialog

I was able to see the leaks and they should be fixed now

FJBDev replied:
> > I stop for now the review until this is fixed. In the following graph you can see the files which I've not yet reviewed
> 
> That sounds good. I am putting the PR in draft mode until I fix all your suggestions/reviews and it is ready to be reviewed again

I have set the PR back to "ready for review" as I have addressed all your code review coments

wolfgang-ch reviewed (changes requested):
The tag images are a great feature but there are issues

FJBDev commented on bundles/net.tourbook/src/net/tourbook/data/TourData.java:
```diff
@@ -12456,6 +12456,15 @@ public void setWorldPixelForWayPoints(final IntObjectHashMap<Point> worldPositio
       _twpWorldPosition.put(projectionHash + zoomLevel, worldPositions);
    }
 
+   public boolean temperaturesExist() {
```
Because there are several temperatures in this case (average, min and max), I felt like the singular `is` didn't apply and as `areTemperatureValid` doesn't sound good, I decided to use apply [this recommendation](https://stackoverflow.com/a/1566778)
But considering your review, I have renamed to `isTemperatureAvailable`

wolfgang-ch replied:
Programming code is not always readable like a prose. In this case I would prefer to keep all methods together which return a boolean value. It's mostly a compromise between both of these.

FJBDev replied:
Sure, I am also liking consistency across the code.
I think sometimes it's hard to find a guideline because not everything is consistent.

For example, I was using `TourData` today and found very confusing that it's not consistent:

[avgPulse](https://github.com/mytourbook/mytourbook/blob/bd51e10dbdd2a5883e489ecba5ad1fdd3cb20a1c/bundles/net.tourbook/src/net/tourbook/data/TourData.java#L414) is used but [powerAvg ](https://github.com/mytourbook/mytourbook/blob/bd51e10dbdd2a5883e489ecba5ad1fdd3cb20a1c/bundles/net.tourbook/src/net/tourbook/data/TourData.java#L583)is defined too
I would think it should be either `avgPulse` and `avgPower` or `pulseAvg` and `powerAvg` but not two different and in this case, i don't know which one follows the correct guideline

wolfgang-ch replied:
`avgPulse` is from the beginning of MT, +15 years ago when there were not many fields in `TourData` but now there are so many fields that I try to "categorize" them -> `power_Avg`

Instead of starting a very big renaming action, with all it's possible risks, I mostly keep it as it is. The text keys in `messages.properties` also have 2 different naming rules which I also don't want to change, because it would need a lot of work to change it without any advantages for a normal user.

![image](https://user-images.githubusercontent.com/1283445/218945446-9cba9db7-b18f-4c7e-b882-3b2b0d3b684a.png)

FJBDev replied:
I get that some fields are older and it's not worth the risk to rename them after such a long time.

My point was that it makes it hard at times to know which rule/convention is to adopt at a given time and therefore can induce some mistakes

wolfgang-ch replied:
Sometimes I also have to look on previous namings to decide how to name something new. With all of the invented names in MT I cannot remember all of them.

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tag/Dialog_TourTag.java:
```diff
@@ -124,24 +159,68 @@ private void createUI(final Composite parent) {
+            label.setText(Messages.Dialog_TourTag_Label_Image);
+            GridDataFactory.fillDefaults().align(SWT.FILL, SWT.BEGINNING).applyTo(label);
 
+            _canvasTagImage = new Label(container, SWT.TRANSPARENCY_ALPHA);
```
I can't reproduce the issue locally. Are you sure that your image original is not horizontal and that maybe Windows Explorer is rotating it in the windows dialog ?
Also, [my code ](https://github.com/FJBDev/mytourbook/blob/6c7e846fac19ef4a4788f6c39b96df4277f0743a/bundles/net.tourbook.photo/src/net/tourbook/photo/ImageUtils.java#L332)never rotates any image

![image](https://user-images.githubusercontent.com/25235665/218577894-4bc9e5d6-6bb0-48f4-82cf-fd4de18d8c06.png)

wolfgang-ch replied:
Do you respect the image orientation?

![image](https://user-images.githubusercontent.com/1283445/218657312-c5e442af-f315-421c-b357-150418b8f3f9.png)

In the MT photo feature, photos are rotated according to the orientation

![image](https://user-images.githubusercontent.com/1283445/218657466-cdd9a5fa-2cfd-418f-bab4-fd8bf4c4c7a3.png)

FJBDev replied:
> Do you respect the image orientation?

I believe so yes because I keep the orientation unchanged, at least that's what I see [in my code.](https://github.com/FJBDev/mytourbook/blob/6c7e846fac19ef4a4788f6c39b96df4277f0743a/bundles/net.tourbook/src/net/tourbook/ui/UI.java#L890)
Should I not be respecting the orientation and decide to rotate the image if needed ?

wolfgang-ch replied:
My original message "The tag image is rotated" should be the opposite: "The tag image is NOT rotated". 

In your [code](https://github.com/FJBDev/mytourbook/blob/6c7e846fac19ef4a4788f6c39b96df4277f0743a/bundles/net.tourbook/src/net/tourbook/ui/UI.java#L890) I don't see that you read the EXIF orientation (when available) and rotate the image accordingly.

My suggestion

* When an image do not contain EXIF orientation data, do not rotate the image
* When an image do contain EXIF orientation data, then rotate the image accordingly

FJBDev replied:
Ok, that makes more sense now, I will implement that

FJBDev replied:
Fixed 

![image](https://user-images.githubusercontent.com/25235665/219716790-5ad8b31d-cf5f-413a-b96b-0629e0bb25c8.png)

wolfgang-ch replied:
I still see the image as not rotated

![image](https://user-images.githubusercontent.com/1283445/219848282-a47d2407-2e6c-4abd-8d2d-59e518247ee2.png)

The photo gallery is rotating it correctly

![image](https://user-images.githubusercontent.com/1283445/219848224-f853ebf2-254b-48d1-a7be-a8f981de6587.png)

wolfgang-ch replied:
In your code...

![image](https://user-images.githubusercontent.com/1283445/219851100-bc5182d4-9c3c-47bb-852e-5833073064e7.png)

...`rotation` could also be `< 0` but when fixing this, the image is still not rotated as it should. 

Maybe you should use the already implemented image resize/rotation code.

FJBDev replied:
>Maybe you should use the already implemented image resize/rotation code.

I thought of it but it didn't seem that the `photo` code was keeping the image transparency when resizing an image so that's why i wrote new code.

FJBDev replied:
Actually, I reused the already implemented image resize/rotation code and it has fixed the below issues:
- Image rotation
- Image pixelization


The image rotation should be fixed now
![image](https://user-images.githubusercontent.com/25235665/220474240-a0b72a21-4220-4e94-9b12-563c4fc9db93.png)

wolfgang-ch replied:
![image](https://user-images.githubusercontent.com/1283445/220536889-15e7aa96-2c98-40b9-bdc1-727e3c5d10e1.png)

😄

wolfgang-ch replied:
Loading larger photo images are now also much faster

FJBDev commented on bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java:
```diff
@@ -5821,6 +5816,11 @@ private boolean updateDb__1_Design(int currentDbVersion, final SplashManager spl
             currentDbVersion = _dbDesignVersion_New = updateDb_047_To_048(conn, splashManager);
          }
 
+         // 47 -> 48    22.X ??
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/data/TourTag.java:
```diff
@@ -114,6 +115,8 @@ public class TourTag implements Cloneable, Comparable<Object> {
    @Transient
    private long                _createId = 0;
 
+   private String              imageFilePath;
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tag/Dialog_TourTag.java:
```diff
@@ -124,24 +159,68 @@ private void createUI(final Composite parent) {
+            label.setText(Messages.Dialog_TourTag_Label_Image);
+            GridDataFactory.fillDefaults().align(SWT.FILL, SWT.BEGINNING).applyTo(label);
 
+            _canvasTagImage = new Label(container, SWT.TRANSPARENCY_ALPHA);
```
Fixed

![image](https://user-images.githubusercontent.com/25235665/218584939-450b5e7c-1ae6-417a-a2d0-0256fbaabb48.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tag/Dialog_TourTag.java:
```diff
@@ -124,24 +159,68 @@ private void createUI(final Composite parent) {
+            label.setText(Messages.Dialog_TourTag_Label_Image);
+            GridDataFactory.fillDefaults().align(SWT.FILL, SWT.BEGINNING).applyTo(label);
 
+            _canvasTagImage = new Label(container, SWT.TRANSPARENCY_ALPHA);
```
>I see the tag image only in the tag dialog and in the context menu. I would like to see them also in the Tagged Tours view

I did that originally, but the image displayed in the cell is so small that I don't see how useful it will be for users except it they buy a magnifying glass 😄 

![image](https://user-images.githubusercontent.com/25235665/218593237-d1b8aa6e-fb70-4826-92ec-8bf41c6ea074.png)

>The tag image filepath is not visible anywhere, I would like to see them in the tag dialog (maybe in a tooltip) and in the tag view in the pref dialog
I can add it to the tooltip

>It seems that the tag images are slowing down the loading of the tag context menu, this may can be optimized by caching the tag images.

Can you show me an example of what you mean by "caching the image" ?

wolfgang-ch replied:
> > I see the tag image only in the tag dialog and in the context menu. I would like to see them also in the Tagged Tours view
> 
> I did that originally, but the image displayed in the cell is so small that I don't see how useful it will be for users except it they buy a magnifying glass 😄

I bought a magnifying glass a few weeks ago 😄 

I would like to see at least in the Tagging view in the pref dialog which tags are containing tag images. It would be helpful to have a new column which displays the image filepathname

wolfgang-ch replied:
> > It seems that the tag images are slowing down the loading of the tag context menu, this may can be optimized by caching the tag images.
> 
> Can you show me an example of what you mean by "caching the image" ?

### Simple
net.tourbook.preferences.GraphColorPainter._imageCache

### Complex
de.byteholder.geoclipse.map.TileImageCache
net.tourbook.photo.PhotoImageCache

FJBDev replied:
ImageCache implemented. See [here](https://github.com/FJBDev/mytourbook/blob/7b762688222bc2fd35a00ce3d58aa689feee1819/bundles/net.tourbook/src/net/tourbook/tag/TagMenuManager.java#L101).

FJBDev replied:
>I would like to see at least in the Tagging view in the pref dialog which tags are containing tag images. It would be helpful to have a new column which displays the image filepathname

Implemented in the "Tour Tags" and "Tagged Tours" views

![image](https://user-images.githubusercontent.com/25235665/219702390-4309d713-ad60-4d8a-8421-fa601a35cab7.png)

wolfgang-ch replied:
Can you add a image file path name column also in the tagging pref dialog because there you define the tags and would be helpful for me

![image](https://user-images.githubusercontent.com/1283445/219848685-3609f46f-7c62-4287-b5a3-cb6b374d883e.png)

FJBDev replied:
Sure, I will add it

FJBDev replied:
Column added

![image](https://user-images.githubusercontent.com/25235665/220470892-bc93fb45-7a28-48fe-a06b-93d369fddde7.png)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/tag/Dialog_TourTag.java:
```diff
@@ -165,15 +246,88 @@ protected void okPressed() {
+
+         if (!Files.exists(Paths.get(imageFilePath))) {
+
+            _lblInvalidImageError.setText(NLS.bind(
+                  Messages.Dialog_TourTag_Label_ImageNotFound,
+                  imageFilePath));
```
A messages dialog already have an error tracking system, an error can be set with `org.eclipse.jface.dialogs.TitleAreaDialog.setErrorMessage(String)` and is used several times in MT

![image](https://user-images.githubusercontent.com/1283445/218697722-83e1c569-757f-4852-95b9-d3b87b028aee.png)

FJBDev replied:
Good idea, I forgot about this one.
Fixed:
![image](https://user-images.githubusercontent.com/25235665/218874143-a10c1d27-b7cc-453f-8e76-e7b1d7d25db1.png)

wolfgang-ch replied:
#### Enhancement
I think it would look better, when these 2 buttons are aligned to the right side. So the image should grab all free horizontal space but is aligned to the left side

FJBDev replied:
Fixed

![image](https://user-images.githubusercontent.com/25235665/219204907-9be0b58b-8983-4fb0-a3ad-4c1ca7716bce.png)

wolfgang-ch commented:
> **Possible improvements:**
> 
>     * When a large tag image is selected, the displayed image is "pixelized" even though the anti-alias is on (@wolfgang-ch Maybe you had this issue before and you know a possible fix ?)

When an image is displayed in full size then this context menu is available

![image](https://user-images.githubusercontent.com/1283445/219849301-f58afc9b-5510-451c-9f3a-6ba003d49eb3.png)


When `Smooth images` is selected, then the code in the red rectangle is used, Here you can also test between smoothed and not smoothed

![image](https://user-images.githubusercontent.com/1283445/219849372-df194981-abc2-4ddd-b51e-a3b47d964f3a.png)

wolfgang-ch commented:
I'm waiting for the further review until the tag image is rotated correctly

FJBDev commented:
> > **Possible improvements:**
> > ```
> > * When a large tag image is selected, the displayed image is "pixelized" even though the anti-alias is on (@wolfgang-ch Maybe you had this issue before and you know a possible fix ?)
> > ```
> 
> When an image is displayed in full size then this context menu is available
> 
> ![image](https://user-images.githubusercontent.com/1283445/219849301-f58afc9b-5510-451c-9f3a-6ba003d49eb3.png)
> 
> When `Smooth images` is selected, then the code in the red rectangle is used, Here you can also test between smoothed and not smoothed
> 
> ![image](https://user-images.githubusercontent.com/1283445/219849372-df194981-abc2-4ddd-b51e-a3b47d964f3a.png)

I already tried to change the antialias but it didn't seem to do anything but [my code](https://github.com/FJBDev/mytourbook/blob/ac65627434af108ddd84ec07e35463d77c6239e0/bundles/net.tourbook.photo/src/net/tourbook/photo/ImageUtils.java#L342) doesn't have the interpolation setting yet so maybe I can try

FJBDev commented:
> I'm waiting for the further review until the tag image is rotated correctly

OK, putting my PR back to `Draft` until I resolve those issues

FJBDev commented:
> > > **Possible improvements:**
> > > ```
> > > * When a large tag image is selected, the displayed image is "pixelized" even though the anti-alias is on (@wolfgang-ch Maybe you had this issue before and you know a possible fix ?)
> > > ```
> > 
> > 
> > When an image is displayed in full size then this context menu is available
> > ![image](https://user-images.githubusercontent.com/1283445/219849301-f58afc9b-5510-451c-9f3a-6ba003d49eb3.png)
> > When `Smooth images` is selected, then the code in the red rectangle is used, Here you can also test between smoothed and not smoothed
> > ![image](https://user-images.githubusercontent.com/1283445/219849372-df194981-abc2-4ddd-b51e-a3b47d964f3a.png)
> 
> I already tried to change the antialias but it didn't seem to do anything but [my code](https://github.com/FJBDev/mytourbook/blob/ac65627434af108ddd84ec07e35463d77c6239e0/bundles/net.tourbook.photo/src/net/tourbook/photo/ImageUtils.java#L342) doesn't have the interpolation setting yet so maybe I can try

This is solved

**Before**:
![image](https://user-images.githubusercontent.com/25235665/220477432-10762d6b-e275-41ca-acf6-895964dba076.png)

**After:**
![image](https://user-images.githubusercontent.com/25235665/220477577-55bc993c-1ad5-4544-9a01-a47d6b61b6e4.png)

FJBDev commented:
> > I'm waiting for the further review until the tag image is rotated correctly
> 
> OK, putting my PR back to `Draft` until I resolve those issues

Setting back the PR to "ready to be reviewed"

wolfgang-ch commented:
> > > I'm waiting for the further review until the tag image is rotated correctly
> > 
> > 
> > OK, putting my PR back to `Draft` until I resolve those issues
> 
> Setting back the PR to "ready for review"

Good timing, the last few days I was at my parents to fill out their tax forms and had no time for MT

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/tagging/TaggingView.java:
```diff
@@ -1420,6 +1421,31 @@ public void update(final ViewerCell cell) {
+
+               cell.setText(TourDatabase.getTagPropertyValue(
+                     (tagItem).tagId,
+                     "TOUR_TAG_IMAGE_FILE_PATH")); //$NON-NLS-1$
```
This string `"TOUR_TAG_IMAGE_FILE_PATH"` is used in several files and should be replaced with a java "constant", the original code with `"TOUR_TAG_AND_CATEGORY_NOTES"` was not better

FJBDev replied:
Fixed. I used the same approach used in [TableColumnFactory](https://github.com/mytourbook/mytourbook/blob/6fa76700a5a69b11d80a99f9b8d1106acfbb37f7/bundles/net.tourbook/src/net/tourbook/ui/TableColumnFactory.java#L43) and used a Constant ID. I did the same for `"TOUR_TAG_AND_CATEGORY_NOTES"`.
But as I didn't see any ID created in `TreeColumnFactory`, let me know if it's not a good solution.

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/UI.java:
```diff
@@ -1147,6 +1267,43 @@ public static void updateUI_Tags(final TourData tourData, final Label tourTagLab
 
+   public static void updateUI_TagsWithImage(final Set<TourTag> tourTags, final Composite tourTagsComposite, final List<CLabel> tagsLabels) {
+
+      // We dispose the current tags labels
+      tagsLabels.forEach(tagLabel -> {
+
+         if (tagLabel != null) {
+
+            Util.disposeResource(tagLabel.getImage());
+         }
+      });
```
Disposing these images causes...

```
org.eclipse.swt.SWTException: Graphic is disposed
	at org.eclipse.swt.SWT.error(SWT.java:4918)
	at org.eclipse.swt.SWT.error(SWT.java:4833)
	at org.eclipse.swt.SWT.error(SWT.java:4804)
	at org.eclipse.swt.graphics.Image.getBounds(Image.java:1245)
	at org.eclipse.swt.custom.CLabel.getTotalSize(CLabel.java:268)
	at org.eclipse.swt.custom.CLabel.computeSize(CLabel.java:147)
	at org.eclipse.swt.layout.RowLayout.computeSize(RowLayout.java:223)
	at org.eclipse.swt.layout.RowLayout.layoutHorizontal(RowLayout.java:291)
	at org.eclipse.swt.layout.RowLayout.computeSize(RowLayout.java:207)
	at org.eclipse.swt.widgets.Composite.computeSizeInPixels(Composite.java:217)
	at org.eclipse.swt.widgets.Control.computeSize(Control.java:663)
	at org.eclipse.swt.layout.GridData.computeSize(GridData.java:493)
	at org.eclipse.swt.layout.GridLayout.layout(GridLayout.java:225)
```

... when another tour is selected.

When I saw this code then I thought that this exception will occure and it occurred. Because the tag image from the image cache is used in serveral places, these images should not be disposed. Because they are relatively small, there should not be a problem with the image limits, they are similar like the tour type images.

FJBDev replied:
It should be fixed now. I probably forgot to update this code when I added the image cache

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourBlogView.java:
```diff
@@ -410,13 +412,14 @@ private void create_22_BlogHeader(final StringBuilder sb) {
-      String tourTitle = _tourData.getTourTitle();
       String tourDescription = _tourData.getTourDescription();
-      String tourWeather = WeatherUtils.buildWeatherDataString(_tourData, true, true, true);
+      final boolean isDescription = StringUtils.hasContent(tourDescription);
 
-      final boolean isDescription = tourDescription.length() > 0;
-      final boolean isTitle = tourTitle.length() > 0;
-      final boolean isWeather = tourWeather.length() > 0;
+      String tourTitle = _tourData.getTourTitle();
+      final boolean isTitle = StringUtils.hasContent(tourTitle);
+
+      String tourWeather = WeatherUtils.buildWeatherDataString(_tourData, true, true, true);
+      final boolean isWeather = StringUtils.hasContent(tourWeather);
```
The old code was better readable where the `is...` vars are all together

![image](https://user-images.githubusercontent.com/1283445/220578327-64a1d920-473c-4420-a5d1-be2c0c46fe30.png)

FJBDev replied:
I tried that because I found it more readable to have each category together (example: `tourDescription` with `isDescription`) but I reverted that.

wolfgang-ch replied:
My code formatting was similar like in the `enableControls` methods, first collect all `is...` var's and then apply them, e.g.

![image](https://user-images.githubusercontent.com/1283445/220836385-0a579141-9b69-487c-976d-26edf3328694.png)

Sometimes the `is...` vars depend on each of them and then it is better readable for me when they are all together

FJBDev replied:
I will follow this pattern

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/UI.java:
```diff
@@ -547,6 +566,51 @@ private static void disableAllControlsInternal(final Composite container) {
+            + "SUM(tourData.TOURDEVICETIME_RECORDED) AS TOTALRECORDEDTIME" + NEW_LINE //                                                                   //$NON-NLS-1$
+            + "FROM " + TourDatabase.JOINTABLE__TOURDATA__TOURTAG + " jTdataTtag" + NEW_LINE //                                                                        //$NON-NLS-1$ //$NON-NLS-2$
+            + "INNER JOIN " + TourDatabase.TABLE_TOUR_DATA + NEW_LINE //                                                       //$NON-NLS-1$
+            + "ON jTdataTtag.TOURDATA_TOURID = tourData.TOURID" + NEW_LINE //                                                                       //$NON-NLS-1$
+            + "GROUP BY jTdataTtag.TOURTAG_TAGID"; //$NON-NLS-1$
+
+      final Map<Long, String> tourTagsAccumulatedValues = new HashMap<>();
+
+      try (Connection connection = TourDatabase.getInstance().getConnection();
+            final PreparedStatement preparedStatement = connection.prepareStatement(sqlQuery)) {
+
+         final ResultSet result = preparedStatement.executeQuery();
+
+         while (result.next()) {
+
+            final long tourTagId = result.getLong(1);
+            float usedMiles = result.getInt(2);
+            final long usedHours = result.getInt(3);
+
+            final StringBuilder tagAccumulatedValues = new StringBuilder();
+
+            tagAccumulatedValues.append(Math.round(usedHours / 3600f));
+            tagAccumulatedValues.append(Messages.tour_editor_label_time_unit);
+
+            tagAccumulatedValues.append(UI.NEW_LINE);
+
+            usedMiles = usedMiles / 1000 / net.tourbook.common.UI.UNIT_VALUE_DISTANCE;
+            tagAccumulatedValues.append(Math.round(usedMiles));
+            tagAccumulatedValues.append(net.tourbook.common.UI.UNIT_LABEL_DISTANCE);
+
+            tourTagsAccumulatedValues.put(tourTagId, tagAccumulatedValues.toString());
+         }
+
+      } catch (final SQLException e) {
+
+         SQL.showException(e, sqlQuery);
+      }
+      return tourTagsAccumulatedValues;
+   }
+
```
I think the `TagManager` would be a better place for this tag code, there is already other SQL code

wolfgang-ch reviewed (changes requested):
It seems that image resources are not your friends, either they are disposed or otherwise they are not disposed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/TourInfoUI.java:
```diff
@@ -1743,15 +1743,15 @@ private void updateUI() {
-                  UI.UNIT_LABEL_SPEED,
-                  IWeather.windSpeedTextShort[getWindSpeedTextIndex(windSpeed)]));
+      final int windSpeed = (int) (_tourData.getWeather_Wind_Speed() / UI.UNIT_VALUE_DISTANCE);
+      if (windSpeed > 0) {
+         _lblWindSpeed.setText(Integer.toString(windSpeed));
+         _lblWindSpeedUnit.setText(
+               String.format(
+                     Messages.Tour_Tooltip_Format_WindSpeedUnit,
+                     UI.UNIT_LABEL_SPEED,
+                     IWeather.windSpeedTextShort[getWindSpeedTextIndex(windSpeed)]));
+      }
```
It is not displayed if the wind direction is not selected
Because if the wind direction is selected to any values other than "empty", then realistically, the wind speed should not be 0
But I can add this in the `if` test

![image](https://user-images.githubusercontent.com/25235665/220688064-24fc2e87-f0d4-4a1f-9084-90a25b5b01d9.png)

FJBDev replied:
Fixed

wolfgang-ch replied:
It looks now better

![image](https://user-images.githubusercontent.com/1283445/220834540-d9ec0c80-7ca0-41dd-bb7c-58591185e1fa.png)

FJBDev replied:
> It looks now better
> 
> ![image](https://user-images.githubusercontent.com/1283445/220834540-d9ec0c80-7ca0-41dd-bb7c-58591185e1fa.png)

Good !

Though, looking at the tooltip with all the weather information, it seems "ugly" to me that the weather info is not clear and "straight" like the info above. But I am not sure of the best solution ... thoughts ?

![image](https://user-images.githubusercontent.com/25235665/221038280-2d349d77-15ac-4cf4-8506-9a8069fcbef6.png)

wolfgang-ch replied:
I remember that the value formatting of this part is not simple. I cannot remember the details why it is now formatted this way but as you can see in the code below just for the temperature value, it is not just a oneliner

![image](https://user-images.githubusercontent.com/1283445/221106671-1e78ca22-6369-4ede-96dd-c6fc605ec80f.png)

![image](https://user-images.githubusercontent.com/1283445/221106787-521b283f-0ced-490a-860c-4188fda823d9.png)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tag/TagMenuManager.java:
```diff
@@ -331,6 +344,29 @@ static void enableRecentTagActions(final boolean isAddTagEnabled, final Set<Long
+    */
+   public static Image getTagImage(final String imageFilePath) {
+
+      var tagImage = TagMenuManager._tagImagesCache.get(imageFilePath);
+
+      if (tagImage == null) {
+
+         tagImage = UI.prepareTagImage(imageFilePath);
+         if (tagImage != null) {
+            TagMenuManager._tagImagesCache.put(imageFilePath, tagImage);
```
Removed

FJBDev replied:
I actually wish Eclipse had a feature to detect those unnecessary prefixes. I know Visual Studio does

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tag/TagMenuManager.java:
```diff
@@ -51,61 +52,64 @@
+   private static boolean                  _isAdvMenu;
+   private static ActionRecentTag[]        _actionsRecentTags;
 
+   private static ActionAllPreviousTags    _actionAllPreviousTags;
```
Probably another random Eclipse automated changed. Reverted

wolfgang-ch replied:
I discovered when it is happening https://github.com/mytourbook/mytourbook/discussions/918#discussion-4340787 when static fields are move before none static fields

FJBDev replied:
Ah ok. In this case, I had to revert manually as Ctrl+Z didn't seem to work

wolfgang-ch replied:
Ctrl+Z works only when you discover the reformatting immediately

FJBDev replied:
Yes, unfortunately, that means that if it's not discovered immediately, then the revert has to be done manually

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/DialogJoinTours.java:
```diff
@@ -170,43 +177,53 @@ public class DialogJoinTours extends TitleAreaDialog implements ITourProvider2 {
+   private Text      _txtTourTitle;
 
-   public DialogJoinTours(final Shell parentShell, final List<TourData> selectedTours) {
+   ITourProvider     _tourProvider;
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/UI.java:
```diff
@@ -830,6 +894,62 @@ public static String getTourTypeLabel(final long tourTypeId) {
+         return null;
+      }
+
+      final Image image = new Image(Display.getDefault(), imageFilePath);
```
Good catch, fixed

FJBDev commented on bundles/net.tourbook.photo/src/net/tourbook/photo/ImageUtils.java:
```diff
@@ -215,16 +207,41 @@ private static Image resize(final Display display, final Image image, final int
+               destData.alphaData[d] = origData.alphaData[o];
+            }
+         }
+         scaledImage = new Image(display, destData);
```
Fixed

FJBDev commented:
> It seems that image resources are not your friends, either they are disposed or otherwise they are not disposed

I am learning a lot though, thanks for your thorough reviews.
I should have addressed all your comments and I have used Sleak and didn't find any non disposed images but let me know if you see otherwise.

FJBDev commented:
On a side note, when I used the existing [Util.disposeResource()](https://github.com/mytourbook/mytourbook/blob/6fa76700a5a69b11d80a99f9b8d1106acfbb37f7/bundles/net.tourbook.common/src/net/tourbook/common/util/Util.java#L654), I noticed that MT has defined the same code **9 times !**

- 4 times [here](https://github.com/mytourbook/mytourbook/blob/6fa76700a5a69b11d80a99f9b8d1106acfbb37f7/bundles/net.tourbook.common/src/net/tourbook/common/UI.java#L1366) even though `Cursor`, `Color`, `Image`, `Font` all inherit from `Resource`, the same function is redefined 4 times

- 3 times [here](https://github.com/mytourbook/mytourbook/blob/6fa76700a5a69b11d80a99f9b8d1106acfbb37f7/bundles/net.tourbook.chart/src/net/tourbook/chart/Util.java#L108) even though `Cursor`, `Color`, `Image` all inherit from `Resource`, the same function is redefined 3 times

 - 1 time [here](https://github.com/mytourbook/mytourbook/blob/6fa76700a5a69b11d80a99f9b8d1106acfbb37f7/bundles/net.tourbook/src/de/byteholder/geoclipse/map/Map2.java#L1175)

Do you agree that we should consolidate them to a single function (I would think that keeping the one in [Util](https://github.com/mytourbook/mytourbook/blob/6fa76700a5a69b11d80a99f9b8d1106acfbb37f7/bundles/net.tourbook.common/src/net/tourbook/common/util/Util.java#L654) would be the logical way to go) ? If yes, do you want me to do it in this PR ? or in a separate one ?

wolfgang-ch commented:
> I have used Sleak and didn't find any non disposed images but let me know if you see otherwise.

I discovered the non disposed images with sleak. I press "Snap & Diff" **before** opening and **after** closing a view/dialog, then you can see the differences. It is helpfull to have a clean workspace with no other opened views and perspectives.

wolfgang-ch commented:
> I noticed that MT has defined the same code **9 times !** 😮

This has historical reasons, I consolidated them now, see commit https://github.com/mytourbook/mytourbook/commit/b32d9b92727df305c7478b677281914d08ea1bd2

wolfgang-ch commented:
> Do you agree that we should consolidate them to a single function

Now there are 4 methods, they have different return values

![image](https://user-images.githubusercontent.com/1283445/220852157-8cbe4b92-15ca-4793-8720-6b8b4ece6f13.png)

FJBDev commented:
> > Do you agree that we should consolidate them to a single function
> 
> Now there are 4 methods, they have different return values
> 
> ![image](https://user-images.githubusercontent.com/1283445/220852157-8cbe4b92-15ca-4793-8720-6b8b4ece6f13.png)

It's a good improvement as it will make it clearer with less duplicated code but you could only have 1 since all the types (`Cursor`, `Image` etc...) inherit from [Resource](https://help.eclipse.org/latest/index.jsp?topic=/org.eclipse.platform.doc.isv/reference/api/org/eclipse/swt/browser/package-summary.html)

wolfgang-ch commented:
> you could only have 1 since all the types (`Cursor`, `Image` etc...) inherit from [Resource](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/reference/api/org/eclipse/swt/graphics/Resource.html)

but then you have to cast the return value which is the reason why there are now 4 instead of 1 method

![image](https://user-images.githubusercontent.com/1283445/221100205-4a58d955-6f4a-4c2e-98b3-2d727b3b82b1.png)
